src/process/index/Indexer.h

```diff
@@ -0,0 +1,202 @@
+     */
+    Indexer(const shared_ptr<Source> &fastqfile,
+            const shared_ptr<Sink> &index,
+            const shared_ptr<IndexEntryStorageStrategy> &storageStrategy,
+            bool enableDebugging = false,
+            bool forceOverwrite = false,
+            bool forbidWriteFQI = false,
+            bool compressDictionaries = true
+    );
+
+    ~Indexer() override = default;
+
+    void setDictionaryCompression(bool value) {
+        this->compressDictionaries = value;
+    }
+
+    bool fulfillsPremises();
+
+    /**
+     * This will create an IndexHeader instance with INDEXER_VERSION and the size of the used IndexEntry struct.
+     */
+    shared_ptr<IndexHeader> createHeader();
+
+    /**
+     * Start the index creation,
+     * @return true, if everything went fine.
+     */
+    bool createIndex();
+
+    void finalizeProcessingForCurrentBlock(stringstream &currentDecompressedBlock, z_stream *strm);
+
+    bool checkAndPrepareForNextConcatenatedPart();
+
+    void storeLinesOfCurrentBlockForDebugMode(std::stringstream &currentDecompressedBlock);
+
+    /**
+     * Overriden to also pass through (copywise, safe but slow but also only with a few entries and in error cases)
+     * error messages from the used IndexWriter and ZLibHelper instance.
+     * @return Merged vector of the objects error messages + the index writers error messages.
+     */
+    vector<string> getErrorMessages() override;
+
+    long getFoundEntries() { return numberOfFoundEntries; };
+
+    /**
+     * For debugging, works only, when enableDebugging was true on object construction.
+     * @return The IndexHeader, which was created during createIndex()
+     */
+    shared_ptr<IndexHeader> getStoredHeader() { return storedHeader; };
+
+    /**
+     * For debugging, works only, when enableDebugging was true on object construction.
+     * @return The index entries, which were created during the index run.
+     */
+    const vector<shared_ptr<IndexEntryV1>> &getStoredEntries() { return storedEntries; }
+
+    int64_t getNumberOfConcatenatedFiles() { return numberOfConcatenatedFiles; }
+
+    shared_ptr<IndexEntryV1> createIndexEntryFromBlockData(const string &currentBlockString,
+                                                           const vector<string> &lines,
+                                                           int64_t &blockOffsetInRawFile,
+                                                           bool lastBlockEndedWithNewline,
+                                                           bool *currentBlockEndedWithNewLine,
+                                                           u_int32_t *numberOfLinesInBlock);
+
+    void storeDictionaryForEntry(z_stream *strm, const shared_ptr<IndexEntryV1>& entry);
+
+    bool writeIndexEntryIfPossible(shared_ptr<IndexEntryV1> &entry, const vector<string> &lines, bool blockIsEmpty);
+
+    void enableWriteOutOfDecompressedBlocksAndStatistics(const path &location) {
```

Comment:
```suggestion
    void enableWritingDecompressedBlocksAndStatistics(const path &location) {
```
Slightly shorter and more fluent to read/think. Just a suggestion.